En este archivo .ipynb voy a trabajar en la creación y desarrollo del conjunto de datos necesario para crear la pestaña en el dashboard correspondiente a los KPI.
Tengo 2 KPI, y en este archivo me voy a centrar en la creación del primero, la consigna del mismo es la siguiente:
- Reducir en un 10% la tasa de homicidios en siniestros viales de los últimos seis meses, en CABA, en comparación con la tasa de homicidios en siniestros viales del semestre anterior.
Definimos a la tasa de homicidios en siniestros viales como el número de víctimas fatales en accidentes de tránsito por cada 100,000 habitantes en un área geográfica durante un período de tiempo específico. Su fórmula es: (Número de homicidios en siniestros viales / Población total) * 100,000

In [1]:
import pandas as pd
hechos = pd.read_excel("../Datasets/homicidios.xlsx")
hechos

ID  N_VICTIMAS      FECHA  AAAA  MM  DD      HORA  HH  \
0    2016-0001           1 2016-01-01  2016   1   1  04:00:00   4   
1    2016-0002           1 2016-01-02  2016   1   2  01:15:00   1   
2    2016-0003           1 2016-01-03  2016   1   3  07:00:00   7   
3    2016-0004           1 2016-01-10  2016   1  10  00:00:00   0   
4    2016-0005           1 2016-01-21  2016   1  21  05:20:00   5   
..         ...         ...        ...   ...  ..  ..       ...  ..   
691  2021-0093           1 2021-12-13  2021  12  13  17:10:00  17   
692  2021-0094           1 2021-12-20  2021  12  20  01:10:00   1   
693  2021-0095           1 2021-12-30  2021  12  30  00:43:00   0   
694  2021-0096           1 2021-12-15  2021  12  15  10:30:00  10   
695  2021-0097           1 2021-11-18  2021  11  18  06:10:00   6   

                               LUGAR_DEL_HECHO TIPO_DE_CALLE  ...  Altura  \
0    AV PIEDRA BUENA Y AV FERNANDEZ DE LA CRUZ       AVENIDA  ...     NaN   
1            AV GRAL PAZ  Y AV DE LOS CORRALES      GRAL PAZ  ...     NaN   
2                           AV ENTRE RIOS 2034       AVENIDA  ...  2034.0   
3        AV LARRAZABAL Y GRAL VILLEGAS CONRADO       AVENIDA  ...     NaN   
4    AV SAN JUAN Y PRESIDENTE LUIS SAENZ PEÑA        AVENIDA  ...     NaN   
..                                         ...           ...  ...     ...   
691                          AV. RIESTRA Y MOM       AVENIDA  ...     NaN   
692                    AU DELLEPIANE Y LACARRA     AUTOPISTA  ...     NaN   
693                        AV. GAONA Y TERRADA       AVENIDA  ...     NaN   
694                         AV. EVA PERON 4071       AVENIDA  ...  4071.0   
695                    PADRE CARLOS MUJICA 709         CALLE  ...   709.0   

                                   Cruce  \
0    FERNANDEZ DE LA CRUZ, F., GRAL. AV.   
1                    DE LOS CORRALES AV.   
2                                    NaN   
3               VILLEGAS, CONRADO, GRAL.   
4                SAENZ PE?A, LUIS, PRES.   
..                                   ...   
691                                  MOM   
692                          LACARRA AV.   
693                              TERRADA   
694                                  NaN   
695                                  NaN   

                                 Dirección Normalizada COMUNA  \
0    PIEDRA BUENA AV. y FERNANDEZ DE LA CRUZ, F., G...      8   
1                 PAZ, GRAL. AV. y DE LOS CORRALES AV.      9   
2                                  ENTRE RIOS AV. 2034      1   
3            LARRAZABAL AV. y VILLEGAS, CONRADO, GRAL.      8   
4              SAN JUAN AV. y SAENZ PEÑA, LUIS, PRES.      1   
..                                                 ...    ...   
691                                  RIESTRA AV. y MOM      7   
692         DELLEPIANE, LUIS, TTE. GRAL. y LACARRA AV.      9   
693                                GAONA AV. y TERRADA     11   
694                                PERON, EVA AV. 4071      9   
695                            PADRE CARLOS MUGICA 709      1   

                                   XY (CABA)         pos x         pos y  \
0      Point (98896.78238426 93532.43437792)  -58.47533969  -34.68757022   
1      Point (95832.05571093 95505.41641999)  -58.50877521  -34.66977709   
2     Point (106684.29090040 99706.57687843)  -58.39040293  -34.63189362   
3      Point (99840.65224780 94269.16534422)  -58.46503904  -34.68092974   
4    Point (106980.32827929 100752.16915795)  -58.38718297  -34.62246630   
..                                       ...           ...           ...   
691   Point (102728.60090138 98186.24929177)  -58.43353773  -34.64561636   
692    Point (99624.29795829 97569.69801131)  -58.46739825  -34.65117757   
693   Point (99116.45492358 101045.23284826)  -58.47293407  -34.61984745   
694    Point (99324.54463985 97676.26932409)  -58.47066794  -34.65021673   
695  Point (107664.16647795 104708.63962087)  -58.37976155  -34.58679619   

      PARTICIPANTES    VICTIMA    A

Bueno, lo que tengo que hacer con esta tabla es trabajar principalmente con las columnas:

- N_VICTIMAS: Para saber cuantas víctimas hubo en ese hecho particular
- FECHA / AAAA / MM: Para hacer la sumatoria de víctimas que hubo tanto en el primer cómo en el segundo semestre.

Procedo a limpiar el dataset dejando únicamente las columnas que voy a utilizar

In [2]:
hechos = hechos[['N_VICTIMAS', 'FECHA', 'AAAA', 'MM']]

In [3]:
hechos

N_VICTIMAS      FECHA  AAAA  MM
0             1 2016-01-01  2016   1
1             1 2016-01-02  2016   1
2             1 2016-01-03  2016   1
3             1 2016-01-10  2016   1
4             1 2016-01-21  2016   1
..          ...        ...   ...  ..
691           1 2021-12-13  2021  12
692           1 2021-12-20  2021  12
693           1 2021-12-30  2021  12
694           1 2021-12-15  2021  12
695           1 2021-11-18  2021  11

[696 rows x 4 columns]

Breve chequeo de nulos y/o duplicados.

In [4]:
hechos.isnull().sum() # No hay nulos

N_VICTIMAS    0
FECHA         0
AAAA          0
MM            0
dtype: int64

In [5]:
hechos[hechos.duplicated()] # Bueno, lo que me doy cuenta es que, hay duplicados porque por ejemplo, al no tener la fila del ID para distinguirlo, dos personas pueden fallecer el mismo día en accidentes de tráfico distintos, y eso se vería duplicado en el conjunto de datos que tengo, pero realmente son hechos diferentes. 

N_VICTIMAS      FECHA  AAAA  MM
6             1 2016-01-24  2016   1
11            1 2016-02-14  2016   2
14            1 2016-02-17  2016   2
15            1 2016-02-17  2016   2
18            1 2016-02-28  2016   2
..          ...        ...   ...  ..
647           1 2021-06-03  2021   6
651           1 2021-06-20  2021   6
658           1 2021-07-22  2021   7
662           1 2021-08-02  2021   8
664           1 2021-08-06  2021   8

[93 rows x 4 columns]

Yo los datos que necesito son los siguientes:

- Año.
- Semestre.
- Cantidad de víctimas en esos semestres.
- Población de CABA en esos semestres.

Primero, para desarrollar el primer prototipo con el cuál luego escalarlo, voy a determinar la cantidad de víctimas de siniestros víales en el primer semestre de 2016 en CABA

In [6]:
# Primero filtro el Dataframe para sólo tener los datos correspondientes a 2016
Hechos_2016 = hechos[hechos["AAAA"] == 2016]
Hechos_2016

N_VICTIMAS      FECHA  AAAA  MM
0             1 2016-01-01  2016   1
1             1 2016-01-02  2016   1
2             1 2016-01-03  2016   1
3             1 2016-01-10  2016   1
4             1 2016-01-21  2016   1
..          ...        ...   ...  ..
139           1 2016-12-27  2016  12
140           1 2016-12-27  2016  12
141           1 2016-12-28  2016  12
142           1 2016-12-29  2016  12
143           1 2016-12-30  2016  12

[144 rows x 4 columns]

Aparentemente, a primera vista, tengo 144 victimas de siniestros víales en 2016, ahora lo voy a dividir en el primer y segundo semestre:

In [7]:
Hechos_2016_1 = Hechos_2016[Hechos_2016["MM"] <= 6] # Primer semestre
print(f"Cantidad de filas primer semestre: {len(Hechos_2016_1)}")
Hechos_2016_2 = Hechos_2016[Hechos_2016["MM"] > 6] # Segundo semestre
print(f"Cantidad de filas segundo semestre: {len(Hechos_2016_2)}")

Cantidad de filas primer semestre: 64
Cantidad de filas segundo semestre: 80


Bueno, una de las razones por las cuáles dejé la columna N_Victimas era para determinar realmente la cantidad de víctimas, porque por más que en la gran mayoría de filas, el valor correspondiente en esta columna sea 1, no siempre es así, y para obtener el dato real de la cantidad de víctimas, tengo que hacer una sumatoria de esta columna. 

In [8]:
Cantidad_victimas_2016_1 = sum(Hechos_2016_1["N_VICTIMAS"]) # Cantidad de víctimas del primer semestre de 2016
print(f"Cantidad de víctimas durante el primer semestre de 2016: {Cantidad_victimas_2016_1}")

Cantidad_victimas_2016_2 = sum(Hechos_2016_2["N_VICTIMAS"]) # Cantidad de víctimas del segundo semestre de 2016
print(f"Cantidad de víctimas durante el segundo semestre de 2016: {Cantidad_victimas_2016_2}")

Cantidad de víctimas durante el primer semestre de 2016: 65
Cantidad de víctimas durante el segundo semestre de 2016: 81


Con esto vemos claramente reflejado que había una víctima más de las filas originales del Dataframe, y este es el dato más preciso que necesitamos. 

Ahora veamos la manera de hacerlo escalable. 

In [9]:
# Comenzamos con un bucle que itera todos los años disponibles en nuestro dataset
for año in hechos["AAAA"].unique():
    print(año)

2016
2017
2018
2019
2020
2021


Funciona correctamente, ahora, con estos años, vamos a filtrar el dataset de hechos.

In [10]:
cantidad_victimas_por_semestre = [] # Creamos una lista vacía que va a almacenar la cantidad de víctimas en ese semestre
for año in hechos["AAAA"].unique(): # Iteramos por cada año único en nuestro conjunto de datos
    hechos_anuales = hechos[hechos["AAAA"] == año] # Creamos un dataframe filtrado por año
    cantidad_victimas_primer_semestre = sum(hechos_anuales[hechos_anuales["MM"] <= 6]["N_VICTIMAS"]) # Obtenemos el dato de la cantidad de víctimas del primer semestre y lo almacenamos en una variable
    cantidad_victimas_por_semestre.append(cantidad_victimas_primer_semestre) # Agregamos la cantidad de víctimas del primer semestre

    # Repetimos lo mismo pero con el segundo semestre
    cantidad_victimas_segundo_semestre = sum(hechos_anuales[hechos_anuales["MM"] > 6]["N_VICTIMAS"]) # Obtenemos el dato de la cantidad de víctimas del primer semestre y lo almacenamos en una variable
    cantidad_victimas_por_semestre.append(cantidad_victimas_segundo_semestre)

# Mostramos en pantalla el resultado de la lista previamente creada
print(cantidad_victimas_por_semestre)


[65, 81, 69, 71, 70, 79, 57, 47, 31, 50, 55, 42]


La salida es la esperada, ya que sigue la lógica anteriormente planteada, y además, la longitud de la lista es de 12 números, lo cuál coincide con:

Cantidad de años (6) * Cantidad de semestres por año (2) = 12

Ahora lo que voy a hacer es obtener un estimativo de la población de CABA en cada uno de esos semestres. 

Según [Wikipedia,](https://es.wikipedia.org/wiki/Demograf%C3%ADa_de_la_ciudad_de_Buenos_Aires) "La población de la Ciudad de Buenos Aires censada en octubre de 2010, cuenta con 2.890.151 habitantes". Mientras que, según [Pagina12,](https://www.pagina12.com.ar/520151-censo-2022-cuales-son-los-barrios-mas-poblados-de-la-ciudad-) en 2022 había "3.120.612 habitantes"

In [11]:
# Tomemos estos datos y hagamos algo de magia
poblacion_caba_2010 = 2890151
poblacion_caba_2022 = 3120612

In [14]:
# Para determinar cuanto creció la población en esos 12 años, hacemos una resta.
crecimiento_en_12_años = poblacion_caba_2022 - poblacion_caba_2010
print(f"La población de CABA creció {crecimiento_en_12_años} habitantes")

La población de Caba creció 230461 habitantes


In [18]:
# Para conocer el crecimiento anual promedio, dividiremos este resultado entre 12 y nos quedaremos sólo con la parte entera
crecimiento_anual_promedio = crecimiento_en_12_años // 12
print(f"El crecimiento anual promedio fué de {crecimiento_anual_promedio} habitantes")

El crecimiento anual promedio fué de 19205 habitantes


In [19]:
# Para conocer, en promedio, y de forma muy aproximada (es importante aclarar esto, porque el calculo que estoy haciendo es líneal, mientras que la gráfica del crecimiento poblacional no tiene una forma lineal) cuánto creció la población por semestre, divido entre dos y me quedo sólo con la parte entera.
crecimiento_semestral_promedio = crecimiento_anual_promedio // 2
print(f"El crecimiento poblacional promedio por semestre fue de {crecimiento_semestral_promedio} habitantes")


El crecimiento poblacional promedio por semestre fue de 9602 habitantes


Ahora, con este número de crecimiento poblacional semestral promedio, voy a determinar la población que hubo en 2016.

In [21]:
poblacion_caba_2016 = poblacion_caba_2010 + (crecimiento_anual_promedio * 6)
print(f"La población de CABA al inicio de 2016 era de {poblacion_caba_2016} habitantes")

La población de CABA al inicio de 2016 era de 3005381 habitantes


Ahora, lo que voy a hacer es almacenar en una lista mediante un bucle, todas estas poblaciones, desde el primer semestre de 2016 hasta el segundo semestre del 2021

In [29]:
# Creo una lista que va a almacenar las poblaciones, originalmente sólo va a tener la poblacion de CABA al inicio de 2016 previamente determinada
lista_con_poblaciones = [poblacion_caba_2016]
for semestre in range(12): # En cada una de las 6 iteraciones:
    poblacion_primer_semestre = lista_con_poblaciones[-1] + crecimiento_semestral_promedio # Se calcula la población del primer semestre
    lista_con_poblaciones.append(poblacion_primer_semestre) # Se agrega a la lista
    # Se repite lo mismo con el segundo semestre
    poblacion_segundo_semestre = lista_con_poblaciones[-1] + crecimiento_semestral_promedio

lista_con_poblaciones
    


[3005381,
 3014983,
 3024585,
 3034187,
 3043789,
 3053391,
 3062993,
 3072595,
 3082197,
 3091799,
 3101401,
 3111003,
 3120605]

In [30]:
# Ahora lo que voy a hacer es eliminar el primer elemento de la lista.
lista_con_poblaciones.pop(0)
print(lista_con_poblaciones)

[3014983, 3024585, 3034187, 3043789, 3053391, 3062993, 3072595, 3082197, 3091799, 3101401, 3111003, 3120605]


A continuación, voy a crear un diccionario que almacene los datos de estas listas para posteriormente transformarlo en un dataframe.

In [31]:
año_semestre_homicidios_poblacion = {
    'Año': [2016, 2016, 2017, 2017, 2018, 2018, 2019, 2019, 2020, 2020, 2021, 2021],
    'Semestre': [1, 2, 1, 2, 1, 2, 1, 2, 1, 2, 1, 2],
    'Victimas':cantidad_victimas_por_semestre,
    'Poblacion':lista_con_poblaciones
}
# Creamos el dataframe a partir del diccionario y lo mostramos en pantalla
KPI_1 = pd.DataFrame(año_semestre_homicidios_poblacion)
KPI_1

Año  Semestre  Victimas  Poblacion
0   2016         1        65    3014983
1   2016         2        81    3024585
2   2017         1        69    3034187
3   2017         2        71    3043789
4   2018         1        70    3053391
5   2018         2        79    3062993
6   2019         1        57    3072595
7   2019         2        47    3082197
8   2020         1        31    3091799
9   2020         2        50    3101401
10  2021         1        55    3111003
11  2021         2        42    3120605

Excelente, ahora este dataframe lo voy a exportar cómo un archivo .csv para posteriormente usarlo dentro del dashboard en PowerBI

In [32]:
KPI_1.to_csv("../Datasets/KPI_1.csv")